File: P_2947_USACO_09_MAR_Look_Up_S.py
n = int(input())
ans, h = [0 for _ in range(n)], []
for _ in range(n):
    h.append(int(input()))
s = []
for i in range(n - 1, -1, -1):
    while s and s[-1][0] <= h[i]:
        s.pop()
    if s:
        ans[i] = s[-1][1]
    s.append([h[i], i + 1])
for x in ans:
    print(x)

File: P_2947_USACO_09_MAR_Look_Up_S.cpp
#include <bits/stdc++.h>
using namespace std;

#define ff first
#define ss second
#define int long long

signed main() {
    int n;
    cin >> n;
    vector<int> ans(n), h(n);
    for (int &hi : h) {
        cin >> hi;
    }
    stack<pair<int, int>> s;
    for (int i = n - 1; i >= 0; --i) {
        while (s.size() and s.top().ff <= h[i]) {
            s.pop();
        }
        if (s.size()) {
            ans[i] = s.top().ss;
        }
        s.emplace(h[i], i + 1);
    }
    for (int x : ans) {
        cout << x << endl;
    }
    return 0;
}
